kvstore/badger/badger.go

```diff
@@ -0,0 +1,179 @@
+package badger
+
+import (
+	"github.com/dgraph-io/badger/v2"
+
+	"github.com/iotaledger/hive.go/kvstore"
+)
+
+// KVStore implements the KVStore interface around a BadgerDB instance.
+type badgerStore struct {
+	instance *badger.DB
+	dbPrefix []byte
+}
+
+// New creates a new KVStore with the underlying BadgerDB.
+func New(db *badger.DB) kvstore.KVStore {
+	return &badgerStore{
+		instance: db,
+	}
+}
+
+func (s *badgerStore) WithRealm(realm kvstore.Realm) kvstore.KVStore {
+	return &badgerStore{
+		instance: s.instance,
+		dbPrefix: realm,
+	}
+}
+
+func (s *badgerStore) Realm() []byte {
+	return s.dbPrefix
+}
+
+// builds a key usable for the badger instance using the realm and the given prefixes.
+func (s *badgerStore) buildKeyPrefix(prefixes []kvstore.KeyPrefix) kvstore.KeyPrefix {
+	prefix := s.dbPrefix
+	for _, optionalPrefix := range prefixes {
+		prefix = append(prefix, optionalPrefix...)
+	}
+	return prefix
+}
+
+func (s *badgerStore) Iterate(prefixes []kvstore.KeyPrefix, consumerFunc kvstore.IteratorKeyValueConsumerFunc) error {
+	return s.instance.View(func(txn *badger.Txn) (err error) {
+		iteratorOptions := badger.DefaultIteratorOptions
+		iteratorOptions.Prefix = s.buildKeyPrefix(prefixes)
+		iteratorOptions.PrefetchValues = true
+
+		it := txn.NewIterator(iteratorOptions)
+		defer it.Close()
+		for it.Rewind(); it.Valid(); it.Next() {
+			item := it.Item()
+			value, err := item.ValueCopy(nil)
+			if err != nil {
+				panic(err)
+			}
+			if !consumerFunc(item.KeyCopy(nil)[len(s.dbPrefix):], value) {
+				break
+			}
+		}
+
+		return nil
+	})
+}
+
+func (s *badgerStore) IterateKeys(prefixes []kvstore.KeyPrefix, consumerFunc kvstore.IteratorKeyConsumerFunc) error {
+	return s.instance.View(func(txn *badger.Txn) (err error) {
+		iteratorOptions := badger.DefaultIteratorOptions
+		iteratorOptions.Prefix = s.buildKeyPrefix(prefixes)
+		iteratorOptions.PrefetchValues = false
+
+		it := txn.NewIterator(iteratorOptions)
+		defer it.Close()
+		for it.Rewind(); it.Valid(); it.Next() {
+			if !consumerFunc(it.Item().KeyCopy(nil)[len(s.dbPrefix):]) {
+				break
+			}
+		}
+
+		return nil
+	})
+}
+
+func (s *badgerStore) Clear() error {
+	return s.instance.DropPrefix(s.dbPrefix)
```

Comment:
Drop prefix is dangerous because it blocks and cancels all other ongoing reads/writes to the db.
Maybe we should iterate and delete here?

Reply:
Ah there is the `DeletePrefix` function we could use?